---
layout: post
title: Project002
subtitle: SNP and INDEL calling
cover-img: /assets/post2-img/indel_blue.jpg
thumbnail-img: https://gpigp.github.io/taehyun/assets/post2-img/marmoset.jpg
tags: [SNP, INDEL, Marmoset, Stanford, GATK4, BWA, Project]

published: true
---

Call SNPs and INDELs
============================

*â• docker í™˜ê²½ì—ì„œ ì§„í–‰ë˜ì—ˆìŠµë‹ˆë‹¤.*
*â• docker environment*
<!-- 1----------------------------------------------------------------------------------------------------------------------------- -->
<h2> 1. Reference & Paired-end whole-genome sequencing data </h2>
<br>
<h6> 2020ë…„ Marmoset Reference </h6>
   
```
wget https://hgdownload.soe.ucsc.edu/goldenPath/calJac4/bigZips/calJac4.fa.gz
gzip -d calJac4.fa.gz
```    
<br>
<h6> SRR7050660 & SRR7050661</h6>
   
```
wget https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos2/sra-pub-run-13/SRR7050660/SRR7050660.1
wget https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos2/sra-pub-run-13/SRR7050661/SRR7050661.1
```   
<span style="color: #2D3748; background-color: #fff5b1">2022.04.26.fixed</span>
<br>
<!-- 2----------------------------------------------------------------------------------------------------------------------------- -->
<h2> 2. SRA Toolkit & Fastq-dump </h2>
<br>
SRR7050660.1, SRR7050661.1 ê°ê°ì€ ë¶„ë¦¬ì‹œí‚¬ìˆ˜ ìˆëŠ”ë°, SRA Toolkitì´ í•„ìš”í•©ë‹ˆë‹¤.   
[<u>ì°¸ì¡°</u>](https://github.com/ncbi/sra-tools/wiki/02.-Installing-SRA-Toolkit)[ğŸ‘ˆ](https://github.com/ncbi/sra-tools/wiki/02.-Installing-SRA-Toolkit) 
<br>
<h6> SRA Toolkit íŒŒì¼ ë‹¤ìš´ </h6>
<h6> SRA Toolkit downloads </h6>
```
wget --output-document sratoolkit.tar.gz http://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-ubuntu64.tar.gz
tar -vxzf sratoolkit.tar.gz
```    
<br>
<h6> í™˜ê²½ë³€ìˆ˜ ì„¤ì • </h6>
<h6> Setting Environment Path </h6>
```
export PATH=$PATH:$PWD/sratoolkit.2.11.2-ubuntu64/bin
which fastq-dump
```
*â• which ì´í›„ ì•„ë¬´ê²ƒë„ ì•ˆëœ¬ë‹¤ë©´ ì„¤ì¹˜ê°€ ì•ˆëœ ê²ƒ.*
*â• If there is nothing after $which, it isn't installed*
<br>
<h6> fastq-dump -I --split-files íŒŒì¼ì´ë¦„ </h6>
```
fastq-dump -I --split-files SRR7050660.1
fastq-dump -I --split-files SRR7050661.1
```
<br>
<!-- 3----------------------------------------------------------------------------------------------------------------------------- -->
<h2> 3. 2ê°œì˜ fastqë¥¼ referenceì— ë§¤í•‘ ì‹œí‚¤ê¸° </h2>
<h2> 3. mapping fastq_1, fastq_2 to reference </h2>
<br>
ë§¤í•‘ ì‹œí‚¤ê¸° ìœ„í•´ì„œëŠ” bwaë¥¼ ì„¤ì¹˜í•´ì•¼ í•œë‹¤.   
[<u>ì°¸ì¡°</u>](https://angus.readthedocs.io/en/2013/bwa-tutorial.html)[ğŸ‘ˆ](https://angus.readthedocs.io/en/2013/bwa-tutorial.html)   
<h6> bwa ì„¤ì¹˜ </h6>   
```
cd /mnt
curl -L -o bwa-0.7.5.tar.bz2 http://sourceforge.net/projects/bio-bwa/files/bwa-0.7.5a.tar.bz2/download
tar xvfj bwa-0.7.5a.tar.bz2
cd bwa-0.7.5a
make
cp bwa /usr/local/bin                // í™˜ê²½ ë³€ìˆ˜ ì„¤ì •
cd home/Lab_Project_002              // ì‘ì—…í•  íŒŒì¼ë¡œ ì´ë™
```
<br>
<h6> reference genome ì¸ë±ìŠ¤ ë§Œë“¤ê¸° </h6>   
<h6> making reference genome index </h6>   
```
bwa index carJac4.fa
```
<br>
<h6> reference & fastq íŒŒì¼ ë§¤í•‘ </h6>
<h6> mapping reference & fastq </h6>
```
// Paired-end seqeunce ì´ë¯€ë¡œ bwa memì„ ì‚¬ìš©í•œë‹¤
// bwa mem -t '# of cores' 'ref file location' 'read1 file' 'read2 file' > 'output'
// ëª…ë ¹ì–´ í•œ ì¤„ë‹¹ ì‹œê°„ ì˜¤ë˜ ê±¸ë¦¬ë¯€ë¡œ ê°ê° ë”°ë¡œ ì‹¤í–‰

bwa mem -M -R "@RG\tID:THKIM\tSM:marmoset\tPL:ILLUMINA\tLB:MLB" -t 32 Reference/calJac4.fa SRR7050660.1_1.fastq SRR7050660.1_2.fastq > SRR7050660.sam
bwa mem -M -R "@RG\tID:THKIM\tSM:marmoset\tPL:ILLUMINA\tLB:MLB" -t 32 Reference/calJac4.fa SRR7050661.1_1.fastq SRR7050661.1_2.fastq > SRR7050661.sam
```
<br>

<!-- 4----------------------------------------------------------------------------------------------------------------------------- -->
<h2> 4 bam file sorting and duplicating </h2>

<h6> GATK ë° Java export </h6>
```
export GATK_LOCAL_JAR=/home/Lab_Project_002/gatk-4.2.2.0/gatk-package-4.2.2.0-local.jar
export JAVA_HOME=~/java
export PATH=$PATH:$JAVA_HOME/bin
```
<br>
<h6> .dict ë° .fai íŒŒì¼ì´ ìˆì–´ì•¼ gatkê°€ ë™ì‘ì´ ëœë‹¤. </h6>
<h6>.dict and .fai file is needed to run gatk. </h6>
```
samtools faidx Reference/calJac4.fa
java -jar /downloads/picard.jar CreateSequenceDictionary \ 
      R=Reference/calJac4.fa \ 
      O=Reference/calJac4.dict
```
<br>
<h6> Sort + Mark Duplicates</h6>
```
gatk MarkDuplicatesSpark \
        -I SRR7050660.sam \
        -M dedup_metrics_0.txt \
        -O SRR7050660.bam
gatk MarkDuplicatesSpark \
        -I SRR7050661.sam \
        -M dedup_metrics_1.txt \
        -O SRR7050661.bam
```
<br>
<span style="color: #2D3748; background-color: #fff5b1">2022.05.07.fixed</span>
<br>
<!-- 5----------------------------------------------------------------------------------------------------------------------------- -->
<h2> 5. DeepVariant </h2>
<br>
<h6> run deepvariant <h6>
```
run_deepvariant --model_type=WGS \
      --ref=/home/Lab_Project_002/2nd_Trial/Reference/calJac4.fa \
      --reads=/home/Lab_Project_002/2nd_Trial/SRR7050661.bam \
      --output_vcf=/home/Lab_Project_002/2nd_Trial/output/marmoset_VCF_gq0_qual0 \
      --num_shards=$(nproc) \
      --logging_dir=/home/Lab_Project_002/2nd_Trial/log \
      --dry_run=false \
      --postprocess_variants_extra_args="qual_filter=0, cnn_homref_call_min_gq = 0"   
run_deepvariant --model_type=WGS \
      --ref=/home/Lab_Project_002/2nd_Trial/Reference/calJac4.fa \
      --reads=/home/Lab_Project_002/2nd_Trial/SRR7050661.bam \
      --output_vcf=/home/Lab_Project_002/2nd_Trial/output/marmoset_VCF_gq0_qual0 \
      --num_shards=$(nproc) \
      --logging_dir=/home/Lab_Project_002/2nd_Trial/log \
      --dry_run=false \
      --postprocess_variants_extra_args="qual_filter=0, cnn_homref_call_min_gq = 0"
```
<br>
<h6> repeat above process changing postprocess variants </h6>
```
"qual_filter=0, cnn_homref_call_min_gq = 0"
"qual_filter=0, cnn_homref_call_min_gq = 0"
"qual_filter=0, cnn_homref_call_min_gq = 0"
"qual_filter=0, cnn_homref_call_min_gq = 0"
```
<br>
<h6> evaluate the outputs </h6>
<br>
<span style="color: #2D3748; background-color: #fff5b1">2022.05.08.fixed</span>
<br>


<h2> ì´í›„ ì§„í–‰í• ê²ƒ </h2>
<h6> 1. GQ filter -> how confident this site is heterozygous, homozygous </h6>
<h6> 2. QUAL filter -> how confident this site is what kind of variation </h6>

<!-- ----------------------------------------------------------------------------------------------------------------------------- -->

<!-- 


<h2> 4. Sam file to Bam file </h2>
<br>
<h6> Samtool Install </h6>
```
cd /root
curl -O -L http://sourceforge.net/projects/samtools/files/samtools/0.1.19/samtools-0.1.19.tar.bz2
tar xvfj samtools-0.1.19.tar.bz2
cd samtools-0.1.19
make
```   

{: .box-error}
```
bam_tview_curses.c:41:10: fatal error: curses.h: No such file or directory
#include "curses.h"
         ^~~~~~~~~~
compilation terminated.
Makefile:129: recipe for target 'bam_tview_curses.o' failed
make: *** [bam_tview_curses.o] Error 1 
```

<h6> â• makeì‹œ ìœ„ì™€ ê°™ì€ ì˜¤ë¥˜ê°€ ëœ¬ë‹¤ë©´ libncureses5-devë¥¼ ì„¤ì¹˜í•´ì•¼ í•©ë‹ˆë‹¤. </h6>
<h6> â• After $make if you have an above error, you have to install libncureses5-dev. </h6>

{: .box-note}
```
apt update && \
apt install -y \
libncurses5-dev \
libncursesw5-dev
```

<br>
<h6> After 'make' </h6>
```
cp samtools /usr/local/bin
cd misc/
cp *.pl maq2sam-long maq2sam-short md5fa md5sum-lite wgsim /usr/local/bin/
cd /home/Lab_Project002
```
<br>
<h6> sam íŒŒì¼ì„ bam íŒŒì¼ë¡œ ë§Œë“¤ê¸° </h6>
<h6> sam file to bam file </h6>
```
samtools import Reference/calJav4.fa.fai SRR7050660.sam SRR7050660.bam
samtools import Reference/calJav4.fa.fai SRR7050661.sam SRR7050661.bam
```
<br>
<span style="color: #2D3748; background-color: #fff5b1">2022.04.28.fixed</span>
<br>

<h2> 5. Bam file to VCF file </h2>
<br>
<h6>GATK ì„¤ì¹˜ ë°©ë²•ì€ ì¶”í›„ì— ê¸°ë¡í•˜ê² ìŠµë‹ˆë‹¤.</h6>
<h6>Install sequence for GATK will be recorded later.</h6>

<h6>ref.dict íŒŒì¼ì„ ë§Œë“¤ê¸° ìœ„í•œ picard.jar ì„¤ì¹˜ ë°©ë²•ë„ ì¶”í›„ì— ê¸°ë¡í•˜ê² ìŠµë‹ˆë‹¤.</h6>
<h6>picard.jar also will be recorded later.</h6>

<h6> Sort + Mark Duplicates</h6>
```
samtools sort -@ 32 SRR7050660.bam SRR7050660_sorted.bam
samtools sort -@ 32 SRR7050661.bam SRR7050661_sorted.bam
```
<br>

<br>
<h6> Mark Duplicates </h6>
```
java -jar /downloads/picard.jar MarkDuplicates \
      I=SRR7050660_sorted.bam \
      O=SRR7050660_dup.bam \
      M=SRR7050660_dup_metrics.txt
      
java -jar /downloads/picard.jar MarkDuplicates \
      I=SRR7050661_sorted.bam \
      O=SRR7050661_dup.bam \
      M=SRR7050661_dup_metrics.txt
```
<br>
<h6> RG(read group) information adding </h6>
```
java -jar /downloads/picard.jar AddOrReplaceReadGroups \
    I=SRR7050660_dup.bam \
    O=SRR7050660_final.bam \
    SORT_ORDER=coordinate \
    RGID=marmoset \
    RGLB=bar \
    RGPL=PNU \
    RGSM=MLB \
    CREATE_INDEX=True

java -jar /downloads/picard.jar AddOrReplaceReadGroups \
    I=SRR7050661_dup.bam \
    O=SRR7050661_final.bam \
    SORT_ORDER=coordinate \
    RGID=marmoset \
    RGLB=bar \
    RGPL=PNU \
    RGSM=MLB \
    CREATE_INDEX=True
    
samtools index SRR7050660_final.bam
samtools index SRR7050661_final.bam
```
<br>
<h6> Call Variants -> make VCF file </h6>
```
gatk --java-options "-Xmx16g" HaplotypeCaller  \
   -R Reference/calJac4.fa \
   -I SRR7050660_final.bam \
   -O SRR7050660.vcf
   
gatk --java-options "-Xmx16g" HaplotypeCaller  \
   -R Reference/calJac4.fa \
   -I SRR7050661_final.bam \
   -O SRR7050661.vcf   
```
<br>

<br> 
```
run_deepvariant --model_type=WGS \
  --ref=/home/Lab_Project_002/2nd_Trial/Reference/calJac4.fa \
  --reads=/home/Lab_Project_002/2nd_Trial/SRR7050661.bam \
  --output_vcf=/home/Lab_Project_002/2nd_Trial/output/marmoset_VCF_gq0_qual0 \
  --num_shards=$(nproc) \
  --logging_dir=/home/Lab_Project_002/2nd_Trial/log \
  --dry_run=false \
  --postprocess_variants_extra_args="qual_filter=0, cnn_homref_call_min_gq = 0"

```
-->
